enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,499 @@
+
+
+
+1. MEG traffic always uses the GR to send and receive traffic.
+2. Egress IP on the primary NIC always uses the GR, even in local gateway mode.
+3. Egress Services always use the host kernel for egress routing.
+
+To provide an ingress/egress point for pods on different networks the most simple solution may appear to be to connect them all to a single gateway router. This introduces an issue where now networks are all connected to a single router, and there may be routing happening between networks that were supposed to be isolated from one another. Furthermore in the future, we will want to extend these networks beyond the cluster, and to do that in OVN would require making a single router VRF aware, which adds more complexity into OVN.
+
+The proposal here is to create a GR per network. With this topology, OVN will create a patch port per network to the br-ex bridge. OVN-Kubernetes will be responsible for being VRF/network aware and forwarding packets via flows in br-ex to the right GR. Each per-network GR will only have load balancers configured on it for its network, and only be able to route to pods in its network. The logical topology would look something like this, if we use an example of having a default primary network, a layer 3 primary network, and a layer 2 primary network:
+
+
+
+![VRF Topology](images/VRFs.svg)
+
+
+In the above diagram, each network is assigned a unique conntrack zone and conntrack mark. These are required in order to be able to handle overlapping networks egressing into the same VRF and SNAT’ing to the host IP. More details in the next section.
+
+
+### Shared Gateway Mode
+
+
+#### Pod Egress
+
+On pod egress, the respective GR of that network will handle doing the SNAT to a unique masquerade subnet IP assigned to this network. For example, in the above diagram packets leaving GR-default would be SNAT’ed to 169.254.169.5 in zone 64005. The packet will then enter br-ex, where flows in br-ex will match this packet, and then SNAT the packet to the node IP in zone 0, and apply its CT mark of 5. Finally, the packet will be recirculated back to table 0, where the packet will be CT marked with 1 in zone 64000, and sent out of the physical interface. In OVN-Kubernetes we use zone 64000 to track things from OVN or the host and additionally, we mark packets from OVN with a CT Mark of 1 and packets from the host with 2. Pseudo openflow rules would look like this (assuming node IP of 172.18.0.3):
```

Comment:
So my diagram was wrong here. We do *not* need to use a unique CT zone and mark for the default cluster network. I updated the diagram for my next push. We need to have everything CT in zone 0 because we need the kernel to detect port conflicts, and choose a new source port. Since the host, OVN cluster default network, and now more networks are sharing the same host IP. We need a final SNAT in that zone to ensure we do not have overlapping conntrack entries.

I think in OVNK you can specify in the config the CT zones to use. We will have to provide a config option for this new range as well (as well as document how big that range should be if you want to have a lot of networks). Does that cover it?